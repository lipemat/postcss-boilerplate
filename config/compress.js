const config = require( '../helpers/package-config' );

/**
 * Compress files using Brotli.
 *
 * Generates .br versions of the .min files if the package.json
 * has `brotliFiles` set to `true`.
 */
module.exports = {
	brotli: {
		options: {
			mode: 'brotli',
		},
		expand: true,
		cwd: config.css_folder,
		dest: config.css_folder,
		extDot: 'last',
		src: [ '**/*.min.css' ],
		ext: '.css.br'
	}
};
