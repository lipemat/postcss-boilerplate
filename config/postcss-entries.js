const config = require( '../helpers/package-config' );

/**
 * Entry points to be loaded by Grunt.
 *
 * Checks each provided file for exists and includes if does.
 *
 * @see getEntries
 */
module.exports = {
	main: config.file_name, // Default: "front-end".
	admin: 'admin',
	blocks: 'blocks'
};
