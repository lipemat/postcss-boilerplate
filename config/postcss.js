let config = require("../helpers/package-config");
const FileSystemLoader = require( '../lib/FileSystemLoader' );
const modulesFolder = process.env.NODE_ENV === 'production' ? '_css-modules-json/min/' : '_css-modules-json/';

let compileOptions = {
	map: true,
	processors: [
		require('postcss-import')({
			extension: 'pcss',
			plugins : [
				require( 'postcss-modules' )( {
					generateScopedName : process.env.NODE_ENV === 'production' ? '[hash:base64:5]' : '[folder]--[name]__[local]',
					Loader: FileSystemLoader.default,
					globalModulePaths : [
						new RegExp( '.*?' + config.theme_path.replace( /\//g, '\\\\' ) + 'pcss', "i" ),
						new RegExp( '.*?' + config.theme_path + 'pcss', "i" ),
					],
					/**
					 * Custom output of css modules JSON file to specified location
					 * Also excludes json files from the global pcss files
					 */
					getJSON: function(cssFileName, json) {
						const path          = require('path');
						const fse = require('fs-extra');
						let directory = path.relative( config.theme_path, cssFileName ).replace( /\\/g, '/' ) + '/';
						//exclude global pcss
						if ( directory.substr(0,4) === 'pcss' ){
							return;
						}
						let cssName       = path.basename(cssFileName, '.css');
						let jsonFileName  = config.theme_path.replace( /\\/g, '/' ) + modulesFolder + directory.replace( cssName + '/','') + cssName + '.json';

						/**
						 * We use the Sync method here to fix issues where JSON is not
						 * being generated.
						 *
						 * @since 2019-01-22
						 */
						fse.outputJsonSync(jsonFileName, json);
					}
				} ),
			],
		}),
        require('postcss-custom-media'),
        require('postcss-nested'),
        //@link https://preset-env.cssdb.org/features
        //By default all stage 2 features work unless specified otherwise
        require('postcss-preset-env'),
        require('postcss-color-mod-function'),
        require('@lipemat/css-mqpacker'),
	],
	parser: require('postcss-scss'),
};

let minOptions = Object.assign({}, compileOptions);
minOptions.map = false;
minOptions.processors = [ ...compileOptions.processors ];
minOptions.processors.push( require( 'postcss-clean' )( {
		level: 2
	} )
);


module.exports = {
	toCSS : {
		options : compileOptions,
		files : {
			'<%= pkg.theme_path %><%= pkg.css_folder %><%= pkg.file_name %>.css': '<%= pkg.theme_path %>pcss/<%= pkg.file_name %>.pcss'
		},
	},

	min: {
		options: minOptions,
		files: {
			'<%= pkg.theme_path %><%= pkg.css_folder %><%= pkg.file_name %>.min.css': '<%= pkg.theme_path %>pcss/<%= pkg.file_name %>.pcss'
		},
	},
};
