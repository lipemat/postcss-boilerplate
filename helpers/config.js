const packageConfig = require( './package-config' );
const path = require( 'path' );


/**
 * Get a config for our /config directory merged with any
 * matching configuration from the project directory.
 *
 * For instance if we have a file named config/babel.config.js in our project
 * we will merge the contents with our config/babel.config.js in favor of whatever
 * is specified with the project's file.
 *
 * @param {String} fileName
 * @returns {object}
 */
function getConfig( fileName ) {
	let config = require( '../config/' + fileName );
	try {
		let localConfig = require( path.resolve( packageConfig.workingDirectory + '/config', fileName ) );
		config = {...config, ...localConfig};
	} catch ( e ) {
	}
	return config;
}

module.exports = {
	getConfig: getConfig
};
