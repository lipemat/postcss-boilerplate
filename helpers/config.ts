import fs from 'fs';
import path from 'path';
import browserslist from 'browserslist';

import {getPackageConfig} from './package-config';


/**
 * Get a config for our /config directory merged with any
 * matching configuration from the project directory.
 *
 * For instance if we have a file named config/babel.config.js in our project
 * we will merge the contents with our config/babel.config.js in favor of whatever
 * is specified with the project's file.
 */
export function getConfig( fileName: string ) {
	let config = require( '../config/' + fileName );
	const packageConfig = getPackageConfig();
	try {
		const localConfig = require( path.resolve( packageConfig.workingDirectory + '/config', fileName ) );
		if ( 'function' === typeof localConfig ) {
			config = {...config, ...localConfig( config )};
		} else {
			config = {...config, ...localConfig};
		}
	} catch ( e ) {
	}
	return config;
}

/**
 * Get the browserslist from the current project.
 *
 * - If specified using standard browserslist config, we will use that.
 * - Fallback to WordPress defaults except for "> 1%".
 */
export function getBrowsersList() {
	const projectBrowsersList = browserslist();
	if ( browserslist( browserslist.defaults ) === projectBrowsersList ) {
		const wp = [ ...require( '@wordpress/browserslist-config' ) ];
		wp.push( 'not and_uc 15.5' );
		return wp;
	}
	return projectBrowsersList;
}

/**
 * If browserslist is not specified, we fall back to WordPress defaults.
 *
 * Return false if a browserslist is specified in the current project.
 *
 * @deprecated Use getBrowsersList() instead.
 *
 * @link https://github.com/browserslist/browserslist#config-file
 *
 * @return {false | string[]}
 */
export const getDefaultBrowsersList = () => {
	if ( browserslist( browserslist.defaults ) === browserslist() ) {
		const wp = [ ...require( '@wordpress/browserslist-config' ) ];
		wp.push( 'not and_uc 15.5' );
		return wp;
	}
	return false;
};


/**
 * Provide CSS properties and media queries to all postcss plugins.
 *
 * If a media-queries files exist, automatically load them.
 * If CSS variables exist, automatically load them.
 *
 * 1. pcss/globals/variables.pcss
 * 3. pcss/globals/media-queries.pcss
 */
export function getExternalFiles(): string[] {
	const externalFiles: string[] = [];
	const packageConfig = getPackageConfig();

	[
		path.resolve( packageConfig.theme_path, 'pcss/globals/media-queries.pcss' ),
		path.resolve( packageConfig.theme_path, 'pcss/globals/variables.pcss' ),
	].forEach( possibleFile => {
		if ( fs.existsSync( possibleFile ) ) {
			externalFiles.push( possibleFile );
		}
	} );

	return externalFiles;
}
