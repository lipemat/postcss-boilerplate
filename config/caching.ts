import {EnumModules, getEnumFilePath} from '../helpers/enum-modules';
import {JsonModules} from '../helpers/get-json';
import tinylr from 'tiny-lr';
import type {Environment} from '../helpers/config';
import {getPackageConfig} from '../helpers/package-config';


export type CachingGruntTasks = {
	reset: {}
	reload: {}
	writeModules: {}
}

let lastCssClasses: string[] = [];

export const config: CachingGruntTasks = {
	reset: {},
	reload: {},
	writeModules: {},
};

function isEnvironment( env: string | undefined ): env is Environment {
	return 'development' === env || 'production' === env;
}

/**
 * Use the same interface as the grunt watch task
 * to trigger a reload.
 *
 * @link https://github.com/gruntjs/grunt-contrib-watch?tab=readme-ov-file#rolling-your-own-live-reload
 */
function triggerReload() {
	tinylr.changed( getEnumFilePath( 'development' ), () => {
		console.log( 'CSS Modules updated, reloading browser.' );
	} );
}

/**
 * Cache management for Enums and JSON
 *
 * `caching:reset` - Reset the content of the Enums and JSON files.
 * `caching:reload` - Reload the browser if a CSS modules classname changes.
 * `caching:writeModules` - Write contents if JSON or PHP Enums for CSS modules.
 *
 */
export default function cachingTask<K extends keyof CachingGruntTasks>( task: K, env: Environment | undefined ) {
	if ( 'reset' === task ) {
		EnumModules._resetContent();
		JsonModules._resetContent();
	}

	if ( 'reload' === task && getPackageConfig().cssEnums ) {
		if ( lastCssClasses.sort().toString() === EnumModules.getCssClasses().sort().toString() ) {
			return;
		}
		lastCssClasses = EnumModules.getCssClasses();
		triggerReload();
	}

	if ( 'writeModules' === task && isEnvironment( env ) && getPackageConfig().combinedJson ) {
		JsonModules.flushToDisk( env );
		if ( getPackageConfig().cssEnums ) {
			EnumModules.flushToDisk( env );
		}
	}
}
