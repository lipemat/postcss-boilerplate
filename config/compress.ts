import {getPackageConfig} from '../helpers/package-config';

const folder = getPackageConfig().css_folder;

export type CompressGruntTasks = {
	brotli: Config;
}

type Options = {
	pretty?: boolean;
	createEmptyArchive?: boolean;

}

type OptionsZip = {
	level?: number;
	mode: 'zip' | 'gzip';
}

type OptionsBrotli = {
	mode: 'brotli';
	brotli?: { [ constant: number ]: number };

}

type OptionsTar = {
	mode: 'tar' | 'zip' | 'tgz';
	archive?: string | ( () => string );
}


type Config = {
	options: Options | OptionsZip | OptionsBrotli | OptionsTar;
	expand: boolean;
	cwd: string;
	dest: string;
	extDot: 'first' | 'last';
	src: string[];
	ext: string;
};


/**
 * Compress files using Brotli.
 *
 * Generates .br versions of the .min files if the package.json
 * has `brotliFiles` set to `true`.
 */
const config: CompressGruntTasks = {
	brotli: {
		options: {
			mode: 'brotli',
		},
		expand: true,
		cwd: folder,
		dest: folder,
		extDot: 'last',
		src: [ '**/*.min.css' ],
		ext: '.css.br',
	},
};

module.exports = config;
