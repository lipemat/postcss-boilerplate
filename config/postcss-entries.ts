import {getPackageConfig} from '../helpers/package-config';

export type PostcssEntries = { [ entry: string ]: string };

const entries: PostcssEntries = {
	main: getPackageConfig().file_name, // Default: "front-end".
	admin: 'admin',
	blocks: 'blocks',
};

/**
 * Entry points to be loaded by Grunt.
 *
 * Checks each provided file for exists and includes if it does.
 *
 * @see getEntries
 */
module.exports = entries;
