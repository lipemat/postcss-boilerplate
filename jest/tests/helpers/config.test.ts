import browserslist = require( 'browserslist' );
import {adjustBrowserslist, getExternalFiles} from '../../../helpers/config';

const {getDefaultBrowsersList, getBrowsersList} = require( '../../../helpers/config' );

afterEach( () => {
	delete process.env.BROWSERSLIST;
} );

describe( 'config', () => {
	test( 'adjustBrowserslist', () => {
		expect( adjustBrowserslist( [ 'chrome 71' ] ) ).toEqual( [
			'chrome 71',
			'not op_mini all',
		] );
	} );

	test( 'getDefaultBrowsersList', () => {
		const expectedBrowsers = [ ...require( '@wordpress/browserslist-config' ) ];
		expectedBrowsers.push( 'not op_mini all' );

		expect( getDefaultBrowsersList() ).toEqual( expectedBrowsers );
		expect( getDefaultBrowsersList() ).toEqual( getBrowsersList() );

		process.env.BROWSERSLIST = 'chrome 71';
		expect( getDefaultBrowsersList() ).toEqual( false );
	} );

	test( 'getBrowsersList', () => {
		const expectedBrowsers = [ ...require( '@wordpress/browserslist-config' ) ];
		expectedBrowsers.push( 'not op_mini all' );

		// Check if the browserslist results change, which may explain other failures.
		expect( browserslist( getBrowsersList() ) ).toMatchSnapshot( 'browserslist' );
		expect( expectedBrowsers ).toMatchSnapshot( 'expectedBrowsers' );

		expect( getBrowsersList() ).toEqual( expectedBrowsers );
		expect( getBrowsersList() ).toEqual( getDefaultBrowsersList() );


		// @notice If this fails, we can probably remove the override in favor of default wp.
		const wpDefaultBrowsers = browserslist( require( '@wordpress/browserslist-config' ), {
			env: 'production',
		} );
		// See if we can remove the override in favor of default wp.
		expect( wpDefaultBrowsers.includes( 'and_uc 15.5' ) ).toBe( false );
		expect( wpDefaultBrowsers.includes( 'op_mini all' ) ).toBe( true );
		expect( browserslist( getBrowsersList() ).includes( 'and_uc 15.5' ) ).toBe( false );
		expect( browserslist( getBrowsersList() ).includes( 'op_mini all' ) ).toBe( false );


		process.env.BROWSERSLIST = 'chrome 71';
		expect( getBrowsersList() ).toEqual( [ 'chrome 71' ] );
	} );

	test( 'getExternalFiles', () => {
		const currentPath = process.cwd();
		const expectedFiles = [
			currentPath + '\\jest\\theme\\pcss\\globals\\media-queries.pcss',
			currentPath + '\\jest\\theme\\pcss\\globals\\variables.pcss',
		];
		expect( getExternalFiles() ).toEqual( expectedFiles );
	} );
} );
