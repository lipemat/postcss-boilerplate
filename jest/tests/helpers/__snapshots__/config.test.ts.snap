// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config getBrowsersList: browserslist 1`] = `
[
  "and_chr 123",
  "android 123",
  "chrome 123",
  "chrome 122",
  "chrome 121",
  "chrome 109",
  "edge 123",
  "edge 122",
  "firefox 124",
  "firefox 123",
  "ios_saf 17.4",
  "ios_saf 17.3",
  "ios_saf 16.6-16.7",
  "op_mob 80",
  "opera 109",
  "opera 108",
  "safari 17.4",
  "safari 17.3",
  "samsung 24",
  "samsung 23",
]
`;

exports[`config getBrowsersList: expectedBrowsers 1`] = `
[
  "> 1%",
  "last 1 Android versions",
  "last 1 ChromeAndroid versions",
  "last 2 Chrome versions",
  "last 2 Firefox versions",
  "last 2 Safari versions",
  "last 2 iOS versions",
  "last 2 Edge versions",
  "last 2 Opera versions",
  "not op_mini all",
]
`;
